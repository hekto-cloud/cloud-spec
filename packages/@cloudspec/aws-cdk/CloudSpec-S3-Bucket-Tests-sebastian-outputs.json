{
  "CloudSpec-S3-Bucket-Tests-sebastian": {
    "bucketName": "cloudspec-s3-bucket-tests-sebas-testbucket560b80bc-rnpfty6mif35"
  }
}
