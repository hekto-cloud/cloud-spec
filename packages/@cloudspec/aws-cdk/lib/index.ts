import { App, Stack, CfnOutput, Tags, IAspect, Aspects, CfnResource, CfnDeletionPolicy, RemovalPolicy, Annotations } from 'aws-cdk-lib';
import { IConstruct } from 'constructs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { AwsCdkCli } from '@aws-cdk/cli-lib-alpha';
import { test, beforeAll, describe, vi, afterAll } from 'vitest';
import { HotswapMode } from 'aws-cdk/lib/api/hotswap/common';
import { RequireApproval } from 'aws-cdk-lib/cloud-assembly-schema';
import os from 'os';
import path from 'path';

class CdkTestHelper {
  private async deployStack(app: App, stackName: string): Promise<string> {
    const assembly = app.synth();
    const cli = AwsCdkCli.fromCdkAppDirectory(assembly.directory, { app: assembly.directory });
    const outputsFile = path.join(process.cwd(), `${stackName}-outputs.json`);

    try {
      await cli.deploy({
        stacks: ['*'],
        hotswap: HotswapMode.FALL_BACK,
        outputsFile,
        requireApproval: RequireApproval.NEVER,
      });
      return outputsFile;
    } catch (error) {
      console.error('Failed to deploy stack:', error);
      throw error;
    }
  }

  private async getStackOutputs(outputsFile: string, stackName: string): Promise<{ [key: string]: string }> {
    const outputsModule = await import(outputsFile);
    const outputs = outputsModule.default || outputsModule;

    if (!outputs[stackName]) {
      throw new Error(`Stack outputs not found for stack: ${stackName}`);
    }

    return outputs[stackName];
  }

  async setupAndDeploy(app: App, stackName: string): Promise<{ [key: string]: string }> {
    const outputsFile = await this.deployStack(app, stackName);
    return this.getStackOutputs(outputsFile, stackName);
  }

  async destroyStack(app: App, stackName: string): Promise<void> {
    const assembly = app.synth();
    const cli = AwsCdkCli.fromCdkAppDirectory(assembly.directory, { app: assembly.directory });

    try {
      await cli.destroy({
        stacks: [stackName],
      });
      console.log(`Stack ${stackName} destroyed successfully.`);
    } catch (error) {
      console.error('Failed to destroy stack:', error);
      throw error;
    }
  }
}

// Add this new class
class EphemeralResourcesAspect implements IAspect {
  visit(node: IConstruct): void {
    if (node instanceof CfnResource) {
      if (node.cfnOptions.deletionPolicy === CfnDeletionPolicy.RETAIN) {
        node.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
      }
      if (node.cfnOptions.updateReplacePolicy === CfnDeletionPolicy.RETAIN) {
        node.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.DELETE;
      }
    }

    // Force delete S3 bucket contents
    if (node instanceof Bucket) {
      node.applyRemovalPolicy(RemovalPolicy.DESTROY);
      (node as any).enableAutoDeleteObjects()
      Annotations.of(node).addWarning('This resource is set for automatic deletion, including all its contents.');
    }
  }
}

export function cloudSpec() {
  let app: App;
  let stack: Stack;
  let outputs: { [key: string]: string } = {};
  let stackName: string;

  const setup = (
    createResources: (stack: Stack, setOutputs: (o: { [key: string]: string }) => void) => void,
    timeout: number = 120000
  ) => {
    beforeAll(async ({name}) => {
      app = new App();
      stackName = inferStackName(name);
      stack = new Stack(app, stackName);

      // Add cloudspec tag to all resources in the stack
      Tags.of(stack).add('cloudspec', 'true');

      // Apply the EphemeralResourcesAspect to the stack
      Aspects.of(stack).add(new EphemeralResourcesAspect());

      createResources(stack, (o) => {
        outputs = o;
        Object.entries(o).forEach(([key, value]) => new CfnOutput(stack, key, { value }));
      });

      outputs = await new CdkTestHelper().setupAndDeploy(app, stackName);
    }, timeout);
  };

  const runTest = (name: string, testFn: (outputs: { [key: string]: string }) => Promise<void>, timeout = 600000) => {
    return test(name, () => testFn(outputs), timeout);
  };

  function inferStackName(testName?: string): string {
    const username = os.userInfo().username;
    const prefix = 'CloudSpec';

    if (testName) {
      return `${prefix}-${testName.replace(/\s+/g, '-')}-${username}`;
    }

    // Check if we're inside a describe block
    if (describe.name) {
      // Use the describe block's name as the stack name
      return `${prefix}-${describe.name.replace(/\s+/g, '-')}-${username}`;
    }

    return `${prefix}-DefaultTest-${username}`;
  }

  // Add this new afterAll block
  afterAll(async () => {
    if (process.env.CLOUDSPEC_DESTROY_AFTER_TEST === 'true') {
      console.log('Destroying stack after test...');
      await new CdkTestHelper().destroyStack(app, stackName);
    }
  }, 600000); // 10-minute timeout for stack destruction

  return { setup, test: runTest };
}
