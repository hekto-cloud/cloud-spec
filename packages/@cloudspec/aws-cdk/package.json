{
  "name": "@cloudspec/aws-cdk",
  "version": "0.2.2",
  "description": "",
  "main": "dist/lib/index.js",
  "types": "dist/lib/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc --outDir dist",
    "dev": "tsc --outDir dist --watch",
    "test": "vitest run --coverage",
    "test:destroy": "CLOUDSPEC_DESTROY_AFTER_TEST=true vitest run"
  },
  "repository": {
    "url": "https://github.com/cloudspec-dev/cloudspec.git"
  },
  "keywords": [],
  "author": {
    "name": "Sebastian Korfmann",
    "email": "sebastian@korfmann.net"
  },
  "license": "Apache-2.0",
  "devDependencies": {
    "@types/node": "^18.11.9",
    "@vitest/coverage-v8": "^1.6.0",
    "aws-cdk": "^2.145.0",
    "aws-cdk-lib": "^2.145.0",
    "constructs": "^10.3.0",
    "typescript": "^5.4.5",
    "vitest": "^1.6.0"
  },
  "peerDependencies": {
    "aws-cdk-lib": "^2.0.0",
    "constructs": "^10.0.0",
    "vitest": "^1.6"
  },
  "dependencies": {
    "@aws-cdk/cli-lib-alpha": "2.147.0-alpha.0"
  }
}
