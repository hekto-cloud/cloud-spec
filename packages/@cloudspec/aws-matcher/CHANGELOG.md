# @cloudspec/aws-matcher

## 0.2.2

### Patch Changes

- 587199c: esm compatibility for vitest

## 0.2.1

### Patch Changes

- cf18b11: fix packaging
