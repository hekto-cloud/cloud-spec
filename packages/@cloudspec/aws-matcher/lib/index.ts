import { expect } from 'vitest';
import { S3Client, HeadObjectCommand, GetObjectCommand, PutObjectCommand } from "@aws-sdk/client-s3";
import { SFNClient, StartExecutionCommand, DescribeExecutionCommand } from "@aws-sdk/client-sfn";
import * as diff from 'diff';
import { Upload } from "@aws-sdk/lib-storage";
import { Readable } from "stream";

// Update the type declaration
declare module 'vitest' {
  interface Assertion<T = any> {
    toHaveKey(properties: { key: string }): Promise<void>;
    toCreateObject(properties: { key: string, body: string | Buffer | Readable }): Promise<void>;
    toMatchS3ObjectSnapshot(properties: { key: string }): Promise<void>;
    toCompleteStepFunctionsExecution(payload: any, timeout?: number): Promise<void>;
  }
}

// Define a type for the matcher result
type MatcherResult = {
  pass: boolean;
  message: () => string;
};

const region = process.env.AWS_REGION || 'us-east-1';
const sfnClient = new SFNClient({ region });
const s3Client = new S3Client({ region });

function getStepFunctionsConsoleUrl(executionArn: string): string {
  try {
    const [, , , region] = executionArn.split(':');
    const baseUrl = `https://${region}.console.aws.amazon.com/states/home`;
    const params = new URLSearchParams({ region });
    return `${baseUrl}?${params.toString()}#/v2/executions/details/${executionArn}`;
  } catch (error) {
    console.error('Error parsing execution ARN:', executionArn, error);
    return 'Unable to generate Step Functions console URL';
  }
}

function createColorfulDiff(actual: string, expected: string): string {
  const differences = diff.diffLines(expected, actual);
  let colorfulDiff = '';

  differences.forEach((part) => {
    // Green for additions, red for deletions
    // If the value is unchanged, it will be grey
    const color = part.added ? '\x1b[32m' : part.removed ? '\x1b[31m' : '\x1b[90m';
    const prefix = part.added ? '+' : part.removed ? '-' : ' ';
    const lines = part.value.split('\n').filter(line => line.trim() !== '');
    lines.forEach(line => {
      colorfulDiff += `${color}${prefix} ${line}\x1b[0m\n`;
    });
  });

  return colorfulDiff;
}

// Custom matchers
const customMatchers = {
  async toHaveKey(received: string, properties: { key: string }): Promise<MatcherResult> {
    const { key } = properties;
    const bucketName = received;
    const headCommand = new HeadObjectCommand({ Bucket: bucketName, Key: key });
    try {
      await s3Client.send(headCommand);
      return {
        message: () => `expected ${key} to exist in S3 bucket ${bucketName}`,
        pass: true,
      };
    } catch (error) {
      return {
        message: () => `expected ${key} to exist in S3 bucket ${bucketName}`,
        pass: false,
      };
    }
  },

  async toCreateObject(received: string, properties: { key: string, body: string | Buffer | Readable }): Promise<MatcherResult> {
    const { key, body } = properties;
    const bucketName = received;
    const s3Client = new S3Client({});

    try {
      const upload = new Upload({
        client: s3Client,
        params: {
          Bucket: bucketName,
          Key: key,
          Body: body,
        },
      });

      await upload.done();

      return {
        pass: true,
        message: () => `Successfully created object with key ${key} in bucket ${bucketName}`,
      };
    } catch (error) {
      return {
        pass: false,
        message: () => `Failed to create object with key ${key} in bucket ${bucketName}: ${error}`,
      };
    }
  },

  async toMatchS3ObjectSnapshot(received: string, properties: { key: string }): Promise<MatcherResult> {
    const { key } = properties;
    const bucketName = received;
    const getCommand = new GetObjectCommand({ Bucket: bucketName, Key: key });
    try {
      const response = await s3Client.send(getCommand);
      const content = await response.Body?.transformToString();

      if (content === undefined) {
        return {
          message: () => `failed to retrieve content from ${key} in S3 bucket ${bucketName}`,
          pass: false,
        };
      }

      // Use Vitest's expect().toMatchSnapshot() for comparison
      try {
        expect(content).toMatchSnapshot();
        return {
          message: () => `expected ${key} content to match snapshot in S3 bucket ${bucketName}`,
          pass: true,
        };
      } catch (error: any) {
        // Extract the diff from the error message
        const actualContent = error.actual;
        const expectedContent = error.expected;

        // Create a colorful diff
        const colorfulDiff = createColorfulDiff(actualContent, expectedContent);

        return {
          message: () =>
            `Snapshot for ${key} in S3 bucket ${bucketName} did not match.\n\nDiff:\n${colorfulDiff}`,
          pass: false,
        };
      }
    } catch (error) {
      return {
        message: () => `failed to retrieve ${key} from S3 bucket ${bucketName}: ${error}`,
        pass: false,
      };
    }
  },

  async toCompleteStepFunctionsExecution(received: string, payload: any, timeout: number = 60000): Promise<MatcherResult & { result: any | null }> {
    const stateMachineArn = received;
    const startTime = Date.now();

    // Start the execution
    const startCommand = new StartExecutionCommand({
      stateMachineArn: stateMachineArn,
      input: JSON.stringify(payload)
    });

    let executionArn: string;
    try {
      const startResponse = await sfnClient.send(startCommand);
      executionArn = startResponse.executionArn!;
    } catch (error) {
      return {
        message: () => `Failed to start Step Functions execution: ${error}`,
        pass: false,
        result: null,
      };
    }

    console.log(`Step Functions execution URL: ${getStepFunctionsConsoleUrl(executionArn)}`);

    while (true) {
      const { status, output } = await sfnClient.send(new DescribeExecutionCommand({ executionArn }));

      if (status === 'SUCCEEDED') {
        return {
          message: () => `Step Functions execution completed successfully`,
          pass: true,
          result: JSON.parse(output || '{}'),
        };
      }

      if (['FAILED', 'TIMED_OUT', 'ABORTED'].includes(status!)) {
        return {
          message: () => `Step Functions execution failed with status: ${status}. View details at the URL printed above.`,
          pass: false,
          result: null,
        };
      }

      if (Date.now() - startTime > timeout) {
        return {
          message: () => `Step Functions execution timed out after ${timeout}ms. Current status: ${status}. View details at the URL printed above.`,
          pass: false,
          result: null,
        };
      }

      await new Promise(resolve => setTimeout(resolve, 5000));
    }
  },
};

// Custom assertions
expect.extend(customMatchers);

export { customMatchers };
