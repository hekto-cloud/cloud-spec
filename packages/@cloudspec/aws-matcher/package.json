{
  "name": "@cloudspec/aws-matcher",
  "version": "0.2.1",
  "description": "",
  "main": "dist/lib/index.js",
  "types": "dist/lib/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc --outDir dist",
    "dev": "tsc --outDir dist --watch",
    "test": "vitest run --coverage",
    "test:destroy": "vitest run"
  },
  "repository": {
    "url": "https://github.com/cloudspec-dev/cloudspec.git"
  },
  "keywords": [],
  "author": {
    "name": "Sebastian Korfmann",
    "email": "sebastian@korfmann.net"
  },
  "license": "Apache-2.0",
  "devDependencies": {
    "@aws-sdk/client-s3": "^3.600.0",
    "@aws-sdk/client-sfn": "^3.600.0",
    "@types/diff": "^5.2.1",
    "@types/node": "^18.11.9",
    "typescript": "^5.4.5",
    "vitest": "^1.6.0"
  },
  "peerDependencies": {
    "@aws-sdk/client-s3": "3",
    "@aws-sdk/client-sfn": "3",
    "vitest": "^1.6"
  },
  "dependencies": {
    "diff": "^5.2.0"
  }
}
