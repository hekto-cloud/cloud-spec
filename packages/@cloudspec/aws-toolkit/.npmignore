
# Exclude typescript source and config
*.ts
tsconfig.json

# Include javascript files and typescript declarations
!*.js
!*.d.ts

# Include .jsii and .jsii.gz
!.jsii
!.jsii.gz
vitest.config.js
node_modules
test
.turbo
coverage
*-outputs.json
*.tgz