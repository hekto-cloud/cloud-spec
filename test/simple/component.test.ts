import { cloudSpec } from '@cloudspec/aws-cdk';
import { describe, expect } from 'vitest';
import { Stack } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import '@cloudspec/aws-matcher';

const cloud = cloudSpec();

describe('S3 Bucket Tests', () => {
  cloud.setup((stack: Stack, setOutputs) => {
    const bucket = new Bucket(stack, 'TestBucket');
    setOutputs({
      bucketName: bucket.bucketName,
    });
  });

  cloud.test('bucket should exist', async (outputs) => {
    const { bucketName } = outputs;
    const key = `test-object-${Date.now()}`;
    const object = { bucketName, key };
    await expect(bucketName).not.toHaveKey({ key });
    await expect(bucketName).toCreateObject({ key, body: 'hello' });
    await expect(bucketName).toHaveKey({ key });
    await expect(bucketName).toMatchS3ObjectSnapshot({ key });
  });
})
