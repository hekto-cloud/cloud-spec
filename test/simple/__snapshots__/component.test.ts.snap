// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`S3 Bucket Tests > bucket should exist 1`] = `"hello"`;
